datalake/src/main/java/com/sequenceiq/datalake/service/upgrade/OrderedOSUpgradeRequestProvider.java

```diff
@@ -96,4 +107,33 @@ private void validateThatEveryInstanceIsPresentInTheConfig(Map<String, List<Stri
                     String.format("The following instances are missing from the ordered OS upgrade request: %s", instancesFromInstanceGroups));
         }
     }
+
+    private void addServiceHostGroupsToOrderedOSUpgradeSet(Map<String, List<String>> instanceIdsByInstanceGroup, int order,
+            List<OrderedOSUpgradeSet> osUpgradeByUpgradeSets) {
+        Set<String> instanceIds = null;
+        do {
```

Comment:
For me, it was hard to understand this cycle's goal (and if it works properly) at first glance. Maybe it would be more readable:
```
Set<String> instanceIds = getServicesHostGroupIds(instanceIdsByInstanceGroup);
while (!instanceIds.isEmpty()) {
  osUpgradeByUpgradeSets.add(new OrderedOSUpgradeSet(order++, instanceIds));
  instanceIds = getServicesHostGroupIds(instanceIdsByInstanceGroup);
}